server/src/test/java/org/elasticsearch/search/SearchServiceTests.java

```diff
@@ -2080,27 +2081,47 @@ public void testSupportsParallelCollectionAffectsSlicing() throws Exception {
         }
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/98534")
     public void testIsParallelCollectionSupportedForResults() {
         SearchSourceBuilder searchSourceBuilder = randomBoolean() ? null : new SearchSourceBuilder();
-        if (searchSourceBuilder != null && randomBoolean()) {
-            searchSourceBuilder.aggregation(new TermsAggregationBuilder("terms"));
+        var aggSupportsParallelCollection = new DateRangeAggregationBuilder("dateRange");
+        var aggDoesNotSupportParallelCollection = new TermsAggregationBuilder("terms");
+        var addedAggregation = searchSourceBuilder != null && randomBoolean()
+            ? randomFrom(aggSupportsParallelCollection, aggDoesNotSupportParallelCollection)
+            : null;
+        if (addedAggregation != null) searchSourceBuilder.aggregation(addedAggregation);
+
+        for (var resultsType : ResultsType.values()) {
+            Runnable boilerplateForExhaustiveSwitch = switch (resultsType) {
+                case NONE, FETCH -> () -> assertFalse(
+                    "NONE and FETCH phases do not support parallel collection.",
+                    SearchService.isParallelCollectionSupportedForResults(resultsType, searchSourceBuilder, randomBoolean())
+                );
+                case DFS -> () -> {
+                    assertTrue(
+                        "DFS phase supports parallel collection.",
+                        SearchService.isParallelCollectionSupportedForResults(resultsType, searchSourceBuilder, randomBoolean())
+                    );
+                };
+                case QUERY -> () -> {
+                    assertFalse(
+                        "Parallel collection can be disabled for Query Phase.",
+                        SearchService.isParallelCollectionSupportedForResults(resultsType, searchSourceBuilder, false)
+                    );
+                    if (addedAggregation == aggDoesNotSupportParallelCollection) {
+                        assertFalse(
+                            "Parallel collection disabled for Query phase with unsupported Agg",
+                            SearchService.isParallelCollectionSupportedForResults(resultsType, searchSourceBuilder, true)
+                        );
+                    } else if (addedAggregation == null || addedAggregation == aggSupportsParallelCollection) {
+                        assertTrue(
+                            "Query phase supports Parallel collection when enabled && does not contain unsupported agg.",
+                            SearchService.isParallelCollectionSupportedForResults(resultsType, searchSourceBuilder, true)
+                        );
+                    }
+                };
+            };
+            boilerplateForExhaustiveSwitch.run();
```

Comment:
what do you need the runnable for? You can have you assertions within the switch without it?

Reply:
As you are aware, a switch _expression_ (vs statement) is exhaustive. This is a desirable quality here, as now our compiler can help enforce correctness: you _must_ modify this test if you add to the `ResultsType` enum, otherwise this test won't compile.

I come from a Kotlin background which has more ergonomics for strong-typing, which is were I developed these preferences.   Brian Goetz discusses the current Java limitations here: https://stackoverflow.com/a/66217353. 
It turns out that to make a switch exhaustive, one must  return a value to make it an expression.  Returning `void` won't work, and surprisingly, after playing around, this seemed to me to be the least offensive way to do it.

I may be biased, but I think the benefit of strong typing outweighs the slight readability cost here in test code, since enforcing correctness is the goal.  It's a matter of preference, and I'm happy to be persuaded to loose the guarantee in favor of simplicity if you lean in that direction.

Reply:
I do  see how a switch is beneficial, but I also think that a runnable here does not really fit in well. I normally work around this switch limitation by adding a default case that throws unsupported operation exception. That fails at runtime instead of at compile time, but we don't add new results type often, actually we very unlikely to add one anytime soon. i agree with the importance of failing for possible new values that are not handled appropriately though.

Another way would be to not require a switc, and use an enumset to iterate over all values that are not DFS or QUERY, and check that concurrency is not enabled for them. This would be maybe a little more readable as follows:

```
assertFalse(
                    "NONE and FETCH phases do not support parallel collection.",
                    SearchService.isParallelCollectionSupportedForResults(NONE, searchSourceBuilderOrNull, randomBoolean())
                );
assertFalse(
                    "NONE and FETCH phases do not support parallel collection.",
                    SearchService.isParallelCollectionSupportedForResults(FETCH, searchSourceBuilderOrNull, randomBoolean())
                );
 assertTrue(
                        "DFS phase always supports parallel collection.",
                        SearchService.isParallelCollectionSupportedForResults(DFS, searchSourceBuilderOrNull, randomBoolean())
                   );

//here goes the rest of the assertions that take QUERY explicitly as argument.

```

I am fine with either of the solutions above (with or without switch) but I would like to remove the runnable part, although I see why you have it. It makes the test difficult to follow. Thanks!

Reply:
Great explanation, thanks!